internal/controllers/topology/cluster/patches/inline/json_patch_generator.go

```diff
@@ -93,56 +99,97 @@ func (j *jsonPatchGenerator) Generate(_ context.Context, req *api.GenerateReques
 			if err != nil {
-				errs = append(errs, errors.Wrapf(err, "failed to generate JSON patches for template %s", template.TemplateRef))
+				errs = append(errs, errors.Wrapf(err, "failed to generate JSON patches for item with uid %q", item.UID))
 				continue
 			}
 
 			// Add jsonPatches to the response.
-			resp.Items = append(resp.Items, api.GenerateResponsePatch{
-				TemplateRef: template.TemplateRef,
-				Patch:       *jsonPatches,
-				PatchType:   api.JSONPatchType,
+			resp.Items = append(resp.Items, runtimehooksv1.GeneratePatchesResponseItem{
+				UID:       item.UID,
+				Patch:     jsonPatches,
+				PatchType: runtimehooksv1.JSONPatchType,
 			})
 		}
 	}
 
-	return resp, kerrors.NewAggregate(errs)
+	if err := kerrors.NewAggregate(errs); err != nil {
+		return &runtimehooksv1.GeneratePatchesResponse{
+			Status:  runtimehooksv1.ResponseStatusFailure,
+			Message: err.Error(),
+		}
+	}
+
+	return resp
 }
 
-// templateMatchesSelector returns true if the template matches the selector.
-func templateMatchesSelector(templateRef *api.TemplateRef, selector clusterv1.PatchSelector) bool {
+// toMap converts a list of Variables to a map of JSON (name is the map key).
+func toMap(variables []runtimehooksv1.Variable) map[string]apiextensionsv1.JSON {
+	variablesMap := map[string]apiextensionsv1.JSON{}
+	for i := range variables {
+		variablesMap[variables[i].Name] = variables[i].Value
+	}
+	return variablesMap
+}
+
+// matchesSelector returns true if the GeneratePatchesRequestItem matches the selector.
+func matchesSelector(req *runtimehooksv1.GeneratePatchesRequestItem, templateVariables map[string]apiextensionsv1.JSON, selector clusterv1.PatchSelector) bool {
+	gvk := req.Object.Object.GetObjectKind().GroupVersionKind()
+
 	// Check if the apiVersion and kind are matching.
-	if templateRef.APIVersion != selector.APIVersion {
+	if gvk.GroupVersion().String() != selector.APIVersion {
 		return false
 	}
-	if templateRef.Kind != selector.Kind {
+	if gvk.Kind != selector.Kind {
 		return false
 	}
 
-	// Check if target matches.
-	switch templateRef.TemplateType {
-	case api.InfrastructureClusterTemplateType:
-		// Check if matchSelector.infrastructureCluster is true.
-		return selector.MatchResources.InfrastructureCluster
-	case api.ControlPlaneTemplateType, api.ControlPlaneInfrastructureMachineTemplateType:
-		// Check if matchSelector.controlPlane is true.
-		return selector.MatchResources.ControlPlane
-	case api.MachineDeploymentBootstrapConfigTemplateType, api.MachineDeploymentInfrastructureMachineTemplateType:
-		if selector.MatchResources.MachineDeploymentClass == nil {
-			return false
+	// Check if the request is for an InfrastructureCluster.
+	if selector.MatchResources.InfrastructureCluster {
+		// Cluster.spec.infrastructureRef holds the InfrastructureCluster.
+		if req.HolderReference.Kind == "Cluster" && req.HolderReference.FieldPath == "spec.infrastructureRef" {
```

Comment:
We could. I used the contract wherever possible (today).

The only reason why I didn't do it for those other paths already is because they are not part of any contract. Essentially it comes down to replicating the structure of our CRDs (Cluster/MachineDeployment) in the contract package. We can definitely do it, but I was a bit conflicted as it felt more like part of our regular API vs. our contracts (which are afaik all related to providers).

Reply:
Thanks for context
Let's open an issue to reconsider this, but given this is the only place where we use this for now I'm fine with the current solution